File: minipack-core/src/test/java/org/minipack/core/example/WriteToBuffer.java
/*
 * Copyright 2024 the MiniPack contributors
 * SPDX-License-Identifier: Apache-2.0
 */
package org.minipack.core.example;

import org.junit.jupiter.api.Test;
import org.minipack.core.LeasedByteBuffer;
import org.minipack.core.MessageReader;
import org.minipack.core.MessageWriter;

import java.io.IOException;

import static org.assertj.core.api.Assertions.assertThat;

class WriteToBuffer extends Example {
  // -8<- [start:snippet]
  LeasedByteBuffer write() throws IOException {
    var output = MessageWriter.BufferOutput.of();
    try (var writer = MessageWriter.of(output)) {
      writer.write("Hello, MiniPack!");
      writer.write(42);
    }
    return output.get();
  }
  // -8<- [end:snippet]

  @Test
  void test() throws IOException {
    var buffer = write();
    var reader = MessageReader.of(buffer);
    assertThat(reader.readString()).isEqualTo("Hello, MiniPack!");
    assertThat(reader.readInt()).isEqualTo(42);
  }
}

File: minipack-kotlin/src/test/kotlin/org/minipack/kotlin/example/WriteToBuffer.kt
/*
 * Copyright 2024 the MiniPack contributors
 * SPDX-License-Identifier: Apache-2.0
 */
package org.minipack.kotlin.example

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.minipack.core.LeasedByteBuffer
import org.minipack.core.MessageWriter.BufferOutput
import org.minipack.kotlin.MessageReaders
import org.minipack.kotlin.MessageWriters

class WriteToBuffer : Example() {
  // -8<- [start:snippet]
  fun write(): LeasedByteBuffer {
    val output = BufferOutput.of()
    MessageWriters.of(output).use { writer ->
      writer.write("Hello, MiniPack!")
      writer.write(42)
    }
    return output.get()
  }
  // -8<- [end:snippet]

  @Test
  fun test() {
    val buffer = write()
    val reader = MessageReaders.of(buffer)
    assertThat(reader.readString()).isEqualTo("Hello, MiniPack!")
    assertThat(reader.readInt()).isEqualTo(42)
  }
}
